keps/sig-node/2000-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,895 @@
+the shutdown from proceeding by `InhibitDelayMaxSec` period to execute some
+critical work prior to shutdown/sleep. Inhibitor Locks were introduced to
+[systemd 183](https://lwn.net/Articles/499480/) (released in 2012).
+
+We believe that making use of systemd is a reasonable approach considering
+almost all new popular linux distros are systemd based (RHEL, Google COS,
+Ubuntu, CentOS, Debian, Fedora, Flatcar Linux, see widespread
+[adoption](https://en.wikipedia.org/wiki/Systemd#Adoption)) and [systemd
+183](https://lwn.net/Articles/499480/) (released in 2012) features support for
+inhibitors.
+
+Thanks to @giuseppe for helping with getting systemd inhibitors working!
+
+### Implementation
+
+Introduce a new Kubelet Config setting, `kubeletConfig.ShutdownGracePeriod`,
+defaulting to 0 seconds. Upon kubelet startup,
+
+*   if the setting is greater than 0 seconds
+    *   kubelet will check with dbus current `InhibitDelayMaxSec` to check if
+        `kubeletConfig.ShutdownGracePeriod <=` `InhibitDelayMaxSec`.
+*   if `kubeletConfig.ShutdownGracePeriod` > `InhibitDelayMaxSec`
+    *   Kubelet will attempt to update the InhibitDelayMaxSec setting, by
+        writing a config file to `/etc/systemd/logind.conf.d/kubelet.conf`, and
+        sending a SIGHUP to logind to update the config setting to ensure that
+        the ShutdownGracePeriod from kubelet config is equal to
+        `InhibitDelayMaxSec`.
+
+After updating the `InhibitDelayMaxSec` on the node if needed, Kubelet will
+query the dbus for the final value of `InhibitDelayMaxSec` set on the node and
+treat min(`InhibitDelayMaxSec`, `kubeletConfig.ShutdownGracePeriod`) as the
+allocatable  shutdown grace period, which will be referred to in this KEP as
+`ShutdownGracePeriod`.
+
+Kubelet will register with dbus as a delay systemd inhibitor lock for the
+`ShutdownGracePeriod` for the shutdown event. Kubelet will also register a
+`PrepareForShutdown` signal which will be emitted prior to the shutdown. Upon
+receiving the signal, Kubelet will have additional `ShutdownGracePeriod` time
+before the actual node will initiate the shutdown.
+
+**Handling the shutdown**
+
+Upon a shutdown occurring, Kubelet will gracefully terminate all the pods
+running on the node and update the Ready condition of the node to false with a
+message `Node Shutting Down`, thereby ensuring new workloads will not get
+scheduled to the node.
+
+Since some of the pods running on the node are often critical for the the
+workloads running on a node (e.g. logging pod daemonset, kubeproxy, kubedns)
+etc, we choose to split the pods running on the node into two categories,
+“critical system pods”, and regular pods. Critical system pods should be
+terminated last, because for example, if the logging pod is terminated first,
+logs from the other workloads will not be captured. Critical system pods are
+identified as those that are in the `system-cluster-critical` or
```

Comment:
I missed this when it came out, but this is super concerning to me.  I don't think these priority classes are "special" in that the Kubelet should hardcode their use.  By hardcoding these, we FORCE workloads that interact with the kubelet or other system infra pods to be in these two priority classes, which breaks the orthogonality of scheduling and resource behavior from the kubelet.

Either there needs to be something that selects for priority classes to treat "specially", or these need to be configurable at startup time in the kubelet.  The former is more flexible, the latter may be more acceptable, but would not allow a service provider to allow users to make that orthogonal.

Reply:
@bobbypage while reviewing the KEP (i was reading through it and thinking about the implications of the intersection with grace period when i noticed this), I think this has to be addressed before we go to beta.

Reply:
Hi @smarterclayton 

Thanks so much for providing your feedback and comments. 

@mrunalp and I discussed this topic in length as part of KEP design and we decided to use priority class of  `system-cluster-critical` and `system-node-critical` to separate "core system workloads (e.g logging deamonsets, etc)" vs regular pods and use that information to determine shutdown ordering.

Unfortunately as I'm sure you're aware there is no existing declarative mechanism to describe pod shutdown ordering and as such we decided to use pod priority as a signal instead.  This is similar to for example pod admission and OOM score adjustment which also use `IsCriticalPod()` as a signal.

@mrunalp and I definitely agree this is not perfect, and happy to discuss if you have some alternative ideas on how to improve it and a potentially better signal for us to use to determine pod shutdown order logic instead for beta moving forward.